packages/std/src/math/int512.rs

```diff
@@ -0,0 +1,1350 @@
+
+impl From<Uint256> for Int512 {
+    fn from(val: Uint256) -> Self {
+        let bytes = [[0u8; 32], val.to_be_bytes()].concat();
+
+        Self::from_be_bytes(bytes.try_into().unwrap())
+    }
+}
+
+impl From<Uint128> for Int512 {
+    fn from(val: Uint128) -> Self {
+        val.u128().into()
+    }
+}
+
+impl From<Uint64> for Int512 {
+    fn from(val: Uint64) -> Self {
+        val.u64().into()
+    }
+}
+
+impl From<u128> for Int512 {
+    fn from(val: u128) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<u64> for Int512 {
+    fn from(val: u64) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<u32> for Int512 {
+    fn from(val: u32) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<u16> for Int512 {
+    fn from(val: u16) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<u8> for Int512 {
+    fn from(val: u8) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<i128> for Int512 {
+    fn from(val: i128) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<i64> for Int512 {
+    fn from(val: i64) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<i32> for Int512 {
+    fn from(val: i32) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<i16> for Int512 {
+    fn from(val: i16) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl From<i8> for Int512 {
+    fn from(val: i8) -> Self {
+        Int512(val.into())
+    }
+}
+
+impl TryFrom<&str> for Int512 {
+    type Error = StdError;
+
+    fn try_from(val: &str) -> Result<Self, Self::Error> {
+        Self::from_str(val)
+    }
+}
+
+impl FromStr for Int512 {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match I512::from_str_radix(s, 10) {
+            Ok(u) => Ok(Self(u)),
+            Err(e) => Err(StdError::generic_err(format!("Parsing i512: {}", e))),
+        }
+    }
+}
+
+impl From<Int512> for String {
+    fn from(original: Int512) -> Self {
+        original.to_string()
+    }
+}
+
+impl fmt::Display for Int512 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // The inner type doesn't work as expected with padding, so we
+        // work around that. Remove this code when the upstream padding is fixed.
+        let unpadded = self.0.to_string();
+        let numeric = unpadded.strip_prefix('-').unwrap_or(&unpadded);
+
+        f.pad_integral(self >= &Self::zero(), "", numeric)
+    }
+}
+
+impl Add<Int512> for Int512 {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        Int512(self.0.checked_add(rhs.0).unwrap())
+    }
+}
+
+impl<'a> Add<&'a Int512> for Int512 {
+    type Output = Self;
+
+    fn add(self, rhs: &'a Int512) -> Self {
+        Int512(self.0.checked_add(rhs.0).unwrap())
+    }
+}
+
+impl Sub<Int512> for Int512 {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        Int512(self.0.checked_sub(rhs.0).unwrap())
+    }
+}
+forward_ref_binop!(impl Sub, sub for Int512, Int512);
```

Comment:
Can we also use this for the Add case above?

Reply:
I think so, yes. Good catch